samza-core/src/main/java/org/apache/samza/storage/SideInputTask.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.storage;
+
+import java.util.Collections;
+import java.util.Set;
+import org.apache.samza.checkpoint.OffsetManager;
+import org.apache.samza.container.RunLoopTask;
+import org.apache.samza.container.TaskInstanceMetrics;
+import org.apache.samza.container.TaskName;
+import org.apache.samza.scheduler.EpochTimeScheduler;
+import org.apache.samza.system.IncomingMessageEnvelope;
+import org.apache.samza.system.SystemStreamPartition;
+import org.apache.samza.task.ReadableCoordinator;
+import org.apache.samza.task.TaskCallback;
+import org.apache.samza.task.TaskCallbackFactory;
+
+
+/**
+ * This class encapsulates the processing logic for side input streams. It is executed by {@link org.apache.samza.container.RunLoop}
+ */
+public class SideInputTask implements RunLoopTask {
```

Comment:
RunLoopTask doc says that if async commit is enabled then process and commit could happen in parallel. Since SideInputTask implements this interface, should we think about adding "synchronized" to SideInputTask.process and SideInputTask.commit methods?

From our offline discussion, I understand that these two methods wont be parallel cause the sideInputRunLoop has asycnCommit = false hard coded and it is always single threaded. 

However, purely from the interface guarantee point of view, shouldnt this impl of RunLoopTask guard against async commits?

Reply:
I don't have strong feelings on this. I am leaning toward including synchronized to indicate explicitly that async commit is not yet supported here, regardless of how it is configured within `RunLoop`.

Just realizing this now, but it would also allow us to more exactly mimic the process & commit exclusivity we had before (with the flush and process thread, one task could be flushing while another processes). We would mark async commit as true in this case, and provide a thread pool to `RunLoop` to execute commit operations in.

@mynameborat thoughts?

Reply:
Yeah you are right. Not sure what I was thinking.

I'll mark both `process` and `commit` as synchronized to be more clear - this makes it more clear that concurrent process and commit is not supported with checkpointing.